File: DSA/Array/0238_Product_of_Array_Except_Self/solution.py
# 0238. Product of Array Except Self
# Difficulty : Medium
# Tags       : array, prefix-sum
# Solved on  : 2019-08-13
# LeetCode   : https://leetcode.com/problems/product-of-array-except-self/

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        length=len(nums)
        L=[0]*length
        R=[0]*length
        Output=[0]* length
        L[0]=1
        for i in range(1,length):
            L[i]=L[i-1]*nums[i-1]
        R[length-1]=1
        for i in reversed(range(length-1)):
            R[i]=R[i+1]*nums[i+1]
        
        for i in range(length):
            Output[i]=L[i]*R[i]
        return Output
            
        

File: DSA/Array/0238_Product_of_Array_Except_Self/cpp/solution.cpp
// 0238. Product of Array Except Self
// Difficulty : Medium
// Tags       : array, prefix-sum
// Solved on  : 2019-08-13
// Attempt    : #1
// LeetCode   : https://leetcode.com/problems/product-of-array-except-self/

class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        
         vector<int> output;
        int mul=1;
        int c=0,p=-1;
        for(int i=0;i<nums.size();i++){
            if(nums[i]==0){
                c++;
                p=i;
            }else{
                mul*=nums[i];
            }
            
        }
        if(c>1)
        {
            for(int i=0;i<nums.size();i++){
                output.push_back(0);
            }
        }else if(c==1){
           for(int i=0;i<nums.size();i++){
               if(i==p){
                  
                    output.push_back(mul);
               }else{
                   output.push_back(0);
               }
                
            } 
        }else{
             for(int i=0;i<nums.size();i++){
               
                 output.push_back(mul/nums[i]);
               
                
            } 
        }
        
        return output;
    }
};